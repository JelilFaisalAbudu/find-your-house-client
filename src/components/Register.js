/* eslint-disable no-underscore-dangle */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable react/button-has-type */
import React, { useState, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Redirect } from 'react-router-dom';

/* eslint-disable consistent-return */
import Form from 'react-validation/build/form';
import Input from 'react-validation/build/input';
import CheckButton from 'react-validation/build/button';
import { isEmail } from 'validator';
import history from '../helpers/history';

import { register } from '../redux/actions/auth';

const required = value => {
  if (!value) {
    return (
      <div className="alert alert-danger" role="alert">
        This field is required!
      </div>
    );
  }
};

const validEmail = value => {
  if (!isEmail(value)) {
    return (
      <div className="alert alert-danger" role="alert">
        This is not a valid email.
      </div>
    );
  }
};

const Register = () => {
  const form = useRef();
  const checkBtn = useRef();

  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [passwordConfirm, setPasswordConfirm] = useState('');
  const [loading, setLoading] = useState(false);
  const [successful, setSuccessful] = useState(false);
  const { isLoggedIn } = useSelector(state => state.auth);

  const { message } = useSelector(state => state.message);
  const dispatch = useDispatch();

  const onChangeName = e => {
    const name = e.target.value;
    setName(name);
  };

  const onChangeEmail = e => {
    const email = e.target.value;
    setEmail(email);
  };

  const onChangePassword = e => {
    const password = e.target.value;
    setPassword(password);
  };

  const onChangePasswordConfirm = e => {
    const password = e.target.value;
    setPasswordConfirm(password);
  };

  const handleRegister = e => {
    e.preventDefault();

    setLoading(true);
    setSuccessful(false);

    form.current.validateAll();

    if (checkBtn.current.context._errors.length === 0) {
      dispatch(register(name, email, password, passwordConfirm))
        .then(() => {
          setSuccessful(true);
          history.push('/home');
        })
        .catch(() => {
          setSuccessful(false);
          setLoading(false);
        });
    } else {
      setLoading(false);
    }
  };

  if (isLoggedIn) {
    return <Redirect to="/home" />;
  }

  return (
    <div className="col-md-12">
      <div className="card card-container">
        <img
          src="//ssl.gstatic.com/accounts/ui/avatar_2x.png"
          alt="profile-img"
          className="profile-img-card"
        />

        <Form onSubmit={handleRegister} ref={form}>
          {!successful && (
            <div>
              <div className="form-group">
                <label htmlFor="name">Name</label>
                <Input
                  type="text"
                  className="form-control"
                  name="name"
                  value={name}
                  onChange={onChangeName}
                  validations={[required]}
                />
              </div>

              <div className="form-group">
                <label htmlFor="email">Email</label>
                <Input
                  type="text"
                  className="form-control"
                  name="email"
                  value={email}
                  onChange={onChangeEmail}
                  validations={[required, validEmail]}
                />
              </div>

              <div className="form-group">
                <label htmlFor="password">Password</label>
                <Input
                  type="password"
                  className="form-control"
                  name="password"
                  value={password}
                  onChange={onChangePassword}
                  validations={[required]}
                />
              </div>

              <div className="form-group">
                <label htmlFor="passwordConfirm">Password Confirmation</label>
                <Input
                  type="password"
                  className="form-control"
                  name="passwordConfirm"
                  value={passwordConfirm}
                  onChange={onChangePasswordConfirm}
                  validations={[required]}
                />
              </div>

              <div className="form-group">
                <button className="btn btn-primary btn-block" disabled={loading}>
                  {loading && (
                  <span className="spinner-border spinner-border-sm" />
                  )}
                  <span className="signup-text">Sign Up</span>
                </button>
              </div>
            </div>
          )}

          {message && (
            <div className="form-group">
              <div className={successful ? 'alert alert-success' : 'alert alert-danger'} role="alert">
                {message}
              </div>
            </div>
          )}
          <CheckButton style={{ display: 'none' }} ref={checkBtn} />
        </Form>
      </div>
    </div>
  );
};

export default Register;
