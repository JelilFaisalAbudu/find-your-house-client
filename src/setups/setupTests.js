import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

Enzyme.configure({ adapter: new Adapter() });
