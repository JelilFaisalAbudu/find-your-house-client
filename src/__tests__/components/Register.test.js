import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import '../../setups/setupTests';
import Register from '../../components/Register';
import store from '../../redux/store';

const setup = () => mount(
  <Provider store={store}>
    <Register />
  </Provider>,
);

describe('', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = setup();
  });

  test('renders the Sign-up page without crashing', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('displays the sign-up for user to click', () => {
    const signUpText = wrapper.find('.signup-text').text();
    expect(signUpText).toMatch(/Sign Up/i);
  });
});
