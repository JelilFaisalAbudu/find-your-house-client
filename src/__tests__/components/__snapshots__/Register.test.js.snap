// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` renders the Sign-up page without crashing 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Register>
    <div
      className="col-md-12"
    >
      <div
        className="card card-container"
      >
        <img
          alt="profile-img"
          className="profile-img-card"
          src="//ssl.gstatic.com/accounts/ui/avatar_2x.png"
        />
        <Form(t)
          onSubmit={[Function]}
        >
          <t
            getValues={[Function]}
            hideError={[Function]}
            onSubmit={[Function]}
            showError={[Function]}
            validate={[Function]}
            validateAll={[Function]}
          >
            <form
              onSubmit={[Function]}
            >
              <div>
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="name"
                  >
                    Name
                    <Control(s)
                      className="form-control"
                      name="name"
                      onChange={[Function]}
                      type="text"
                      validations={
                        Array [
                          [Function],
                        ]
                      }
                      value=""
                    >
                      <s
                        className="form-control"
                        error={
                          <div
                            className="alert alert-danger"
                            role="alert"
                          >
                            This field is required!
                          </div>
                        }
                        name="name"
                        onBlur={[Function]}
                        onChange={[Function]}
                        type="text"
                        value=""
                      >
                        <div>
                          <input
                            className="form-control"
                            name="name"
                            onBlur={[Function]}
                            onChange={[Function]}
                            type="text"
                            value=""
                          />
                        </div>
                      </s>
                    </Control(s)>
                  </label>
                </div>
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="email"
                  >
                    Email
                  </label>
                  <Control(s)
                    className="form-control"
                    name="email"
                    onChange={[Function]}
                    type="text"
                    validations={
                      Array [
                        [Function],
                        [Function],
                      ]
                    }
                    value=""
                  >
                    <s
                      className="form-control"
                      error={
                        <div
                          className="alert alert-danger"
                          role="alert"
                        >
                          This field is required!
                        </div>
                      }
                      name="email"
                      onBlur={[Function]}
                      onChange={[Function]}
                      type="text"
                      value=""
                    >
                      <div>
                        <input
                          className="form-control"
                          name="email"
                          onBlur={[Function]}
                          onChange={[Function]}
                          type="text"
                          value=""
                        />
                      </div>
                    </s>
                  </Control(s)>
                </div>
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="password"
                  >
                    Password
                  </label>
                  <Control(s)
                    className="form-control"
                    name="password"
                    onChange={[Function]}
                    type="password"
                    validations={
                      Array [
                        [Function],
                      ]
                    }
                    value=""
                  >
                    <s
                      className="form-control"
                      error={
                        <div
                          className="alert alert-danger"
                          role="alert"
                        >
                          This field is required!
                        </div>
                      }
                      name="password"
                      onBlur={[Function]}
                      onChange={[Function]}
                      type="password"
                      value=""
                    >
                      <div>
                        <input
                          className="form-control"
                          name="password"
                          onBlur={[Function]}
                          onChange={[Function]}
                          type="password"
                          value=""
                        />
                      </div>
                    </s>
                  </Control(s)>
                </div>
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="passwordConfirm"
                  >
                    Password Confirmation
                  </label>
                  <Control(s)
                    className="form-control"
                    name="passwordConfirm"
                    onChange={[Function]}
                    type="password"
                    validations={
                      Array [
                        [Function],
                      ]
                    }
                    value=""
                  >
                    <s
                      className="form-control"
                      error={
                        <div
                          className="alert alert-danger"
                          role="alert"
                        >
                          This field is required!
                        </div>
                      }
                      name="passwordConfirm"
                      onBlur={[Function]}
                      onChange={[Function]}
                      type="password"
                      value=""
                    >
                      <div>
                        <input
                          className="form-control"
                          name="passwordConfirm"
                          onBlur={[Function]}
                          onChange={[Function]}
                          type="password"
                          value=""
                        />
                      </div>
                    </s>
                  </Control(s)>
                </div>
                <div
                  className="form-group"
                >
                  <button
                    className="btn b btn-primary btn-block btn-submit"
                    disabled={false}
                    type="submit"
                  >
                    <span
                      className="signup-text"
                    >
                      Sign Up
                    </span>
                  </button>
                </div>
              </div>
              <Button(s)
                style={
                  Object {
                    "display": "none",
                  }
                }
              >
                <s
                  hasErrors={true}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                >
                  <button
                    disabled={true}
                    style={
                      Object {
                        "display": "none",
                      }
                    }
                  />
                </s>
              </Button(s)>
            </form>
          </t>
        </Form(t)>
      </div>
    </div>
  </Register>
</Provider>
`;
